### DISCUSSION TITLE
Inconsistentencies between Android and iOS

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#678] [General]
I want to excuse myself for this seemingly never ending story but...

I just noticed that the change from `BarcodeReader` to `ZXingCpp` in the Android wrapper that @markusfisch implemented [here](https://github.com/zxing-cpp/zxing-cpp/commit/23ce4aba2923c657ba78851b6ca5b2ea0145b6d8) should have included the changes I just commited [here](https://github.com/zxing-cpp/zxing-cpp/commit/06082d5a54211a39740b3d10e9a74853dac5df4a) to make the transition from the "reader object" to a "global function" like we have in the C++ API complete. I agreed to that change and still can see the benefit of the increased consistency c++ vs java.

But this made me realize that the iOS wrapper still uses the "reader object" model. Now that we put all the effort in to make everything more coherent, that fact does bother me now. When the iOS wrapper was first introduced, I did not question the naming (e.g. `ZXIBarcodeReader`) or the general structure of the API because I thought that might be the most natural and best practice approach for some ObjectiveC code that I can barely read (where did the 'I' in 'ZXI' come from?). After spending the last weeks with you guys to polish the whole wrapper experience I'm now questioning it.

I'm not going to touch the iOS wrapper as I can't even compile it but I'd like to do either of the following:
1) basically revert the Android related refactoring and bring back the `BarcodeReader`
2) do the same "reader object" to "global function" transition on the iOS side

If there must be an inconsistency between c++ / android / ios, then having the two mobile platforms feel the same, makes more sense to me.

Thoughts? @benjohnde @markusfisch @parallaxe

**Comment by axxel** [comment]
After some more thinking about the situation, I realized that having this 'singleton' `object ZXingCpp` in Kotlin is not really the canonical equivalent of the C++ API either. One would simply have a global `fun readBarcodes()`. I was wondering how that translates to Java since everything has to be a member of a class. The answer can be found [here](https://kotlinlang.org/docs/java-to-kotlin-interop.html#package-level-functions). I can't help but feel like this creates way more quirks from a user perspective (who as an Android developer hardly cares about C++) than is worth it by gaining some rather artificial similarity to the C++ API.

If I completely disregard all existing API and history and want to come up with the most straight forward interface then I believe something along the lines of [this](https://developer.mozilla.org/en-US/docs/Web/API/BarcodeDetector) would come out. The current iOS code is certainly closer to that than the current Android code.

Bottom line: to not drag this discussion along even more and waste all your time on something that is actually a pretty trivial API, I've decided to go back to the "reader object" model for Android.

EDIT: ~~I'll stick with `DecodeHints` instead of `Options` and `read` instead of `detect`.~~ Interesting what happens if you start taking a step back and question the status quo: `DecodeHints` is a bad name at least ever since I replaced `MultiFormatReader.decode()` with `ReadBarcode()`. I should have renamed it then and there, maybe to `ReadOptions`. With the option to use nested classes in Kotlin, I believe calling this `BarcodeReader.Options` (instead of `BarcodeReader.ReadOptions`) is the most obvious choice. ObjectiveC does not support nesting classes (says a google search), so there the most obvious choice seems `ReadOptions` or `BarcodeReaderOptions`. Interestingly I saw the `DecodeHintsFromZXIOptions` function in the iOS wrapper :).

@parallaxe / @markusfisch:
 * if `BarcodeReader` is too generic a term for you, you could simply `import zxingcpp.BarcodeReader as ZXBarcodeReader` or whatever.
 * can you explain to me what the reason was behind the `ZXI` prefix in the original iOS wrapper PR?
 * am I correct in interpreting the [ZXIBarcodeReader](https://github.com/zxing-cpp/zxing-cpp/blob/06082d5a54211a39740b3d10e9a74853dac5df4a/wrappers/ios/Sources/Wrapper/Reader/ZXIBarcodeReader.h#L13) API that you could construct such an object with a `hints` parameter and you can access/modify a `hints` property of this object after creation (the demo code does not show this)?

**Comment by markusfisch** [comment]
> basically revert the Android related refactoring and bring back the BarcodeReader

Well, now that you already did it, there's not much to talk about anymore I guess 😉 If you like that approach better, it's probably the right thing to do.

My guess is that adding a `BarcodeWriter` someday will make this setup more complex as you would want to have two native libraries then, but I may as well be wrong about that.

> do the same "reader object" to "global function" transition on the iOS side

I'm not sure this translates to Objective-C. Also I don't think that all wrappers need to have the same architecture. Rather, I would find it more important that all wrappers work with the same symbol names as they are used in the native code, so we all talk about the same things no matter what wrapper we are using.

> can you explain to me what the reason was behind the ZXI prefix in the original iOS wrapper PR?

Frankly, I have no idea 😬 I guess only @parallaxe or @ChristianNorbertBraun can answer this.

> am I correct in interpreting the [ZXIBarcodeReader](https://github.com/zxing-cpp/zxing-cpp/blob/06082d5a54211a39740b3d10e9a74853dac5df4a/wrappers/ios/Sources/Wrapper/Reader/ZXIBarcodeReader.h#L13) API that you could construct such an object with a hints parameter and you can access/modify a hints property of this object after creation (the demo code does not show this)?

Yes, exactly, [initWithHints](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/ios/Sources/Wrapper/Reader/ZXIBarcodeReader.mm#L47) takes a pointer to a `ZXIDecodeHints` structure, that can be accessed and modified at any time.

So `ZXIBarcodeReader` is very close to what `BarcodeReader` is in Android now.

**Reply by axxel** [reply]
> My guess is that adding a `BarcodeWriter` someday will make this setup more complex as you would want to have two native libraries then, but I may as well be wrong about that.

I don't believe it will and even if, it won't matter from a client side perspective and that is what is important from my point of view. Like there is still a `MultiFormatReader` class in the codebase but a C++ client does not see it anymore.

> I'm not sure this translates to Objective-C. Also I don't think that all wrappers need to have the same architecture. Rather, I would find it more important that all wrappers work with the same symbol names as they are used in the native code, so we all talk about the same things no matter what wrapper we are using.

I'd like the used terminology to be consistent as well (see #104). Since I am now convinced and aware of the fact that `DecodeHints` is a bad name and is only still there because it has always been there and I never questioned it, I want to get rid of it (everywhere). Since the wrapper code will break backward compatibility in the next release anyway and the new publishing channels might draw more people, I want to remove that term for the two wrappers before the next release. The c++ API will come later and take longer but will change as well.

I considered `ReadOptions`, `ReadingOptions`, `ReaderOptions` and `BarcodeReaderOptions`. Because I like brevity but `ReadOptions` sounds to much like an action (see `ReadBarcodes`), my current preference is `ReaderOptions` (and `WriterOptions` respectively).

> > can you explain to me what the reason was behind the ZXI prefix in the original iOS wrapper PR?
> 
> Frankly, I have no idea 😬 I guess only @parallaxe or @ChristianNorbertBraun can answer this.

I'm looking forward to that and depending on the answer might want to change or even remove that prefix.

> Yes, exactly, [initWithHints](https://github.com/zxing-cpp/zxing-cpp/blob/master/wrappers/ios/Sources/Wrapper/Reader/ZXIBarcodeReader.mm#L47) takes a pointer to a `ZXIDecodeHints` structure, that can be accessed and modified at any time.

Good. I'd like to change the `ZXIBarcodeWriter` to behave consistently, meaning there is an `options` member of the class and `write()` does not get it as a parameter anymore.

> So `ZXIBarcodeReader` is very close to what `BarcodeReader` is in Android now.

That was my point. :)

If any of you will want to spend more time on this subject of 'naming cosmetic', I'd appreciate your help with the iOS code, otherwise I will throw changes at the CI system until it compiles ;).

**Reply by markusfisch** [reply]
> I considered ReadOptions, ReadingOptions, ReaderOptions and BarcodeReaderOptions. Because I like brevity but ReadOptions sounds to much like an action (see ReadBarcodes), my current preference is ReaderOptions (and WriterOptions respectively).

`ReaderOptions`/`WriterOptions` would also be my favorites, for the same reasons.

> I'd like to change the ZXIBarcodeWriter to behave consistently, meaning there is an options member of the class and write() does not get it as a parameter anymore.

Now that's just consistent 😉 

> If any of you will want to spend more time on this subject of 'naming cosmetic', I'd appreciate your help with the iOS code, otherwise I will throw changes at the CI system until it compiles ;).

Unfortunately, I'm a bit busy this week, but maybe can help with this next week.

**Reply by axxel** [reply]
I went ahead and did the renaming of `Hints` -> `Options` and moved the parameter of the Writer from `write` to `init` in #682 . It took me 4 retries but now it compiles. Comments welcome.

@parallaxe / @ChristianNorbertBraun: I'd appreciate a brief explanation of your thought process behind the `ZXI` prefix. Was a prefix necessary or just a cosmetic decision?

**Reply by axxel** [reply]
@benjohnde I just noticed that your ZXingObjC code base also uses a common prefix ('ZX'). Do you know why? Do you have arguments for or against this type of namespacing in ObjectiveC? If there was a convincing argument for a prefix, I'd prefer ZX over ZXI.

**Reply by axxel** [reply]
@ everybody: it turns out I could have answered those prefix questions by a simple google search myself... As Apple [states](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/Conventions/Conventions.html) in ObjectiveC _every_ user defined class is supposed to have a 3 letter prefix. 2 letter prefixes are reserved for Apples own use. Swift does not need this anymore but since the wrapper is itself written in ObjectiveC, here we are.

With those rules in place I might have chosen ZXC (for ZXingCpp) instead of ZXI but changing that now seems not worth it.

Since I stumbled upon that info: If ObjectiveC was not a requirement, there would be the option of writing a slim native Swift wrapper around the C-API (see [here](https://developer.apple.com/documentation/swift/using-imported-c-functions-in-swift)). Or wait a little longer and then use the upcoming [Swift to C++ bridge](https://www.swift.org/documentation/cxx-interop/#introduction) directly.

Not a question of any relevance at the moment but who of you is actually still depending on ObjectiveC?

**Reply by markusfisch** [reply]
Didn't know that about the three letter prefix, actually 🙈 Well, now that's at least an explanation.

> Not a question of any relevance at the moment but who of you is actually still depending on ObjectiveC?

In general, we are very excited about the fact that Swift will offer more direct and better C++ integration in the future. It's probably just a bit early to jump on board, as only Swift 5.9/Xcode15 are supported.

**Reply by axxel** [reply]
I introduced the term `ReaderOptions` throughout the complete code base. I'm closing this discussion now.
